src/main/java/org/ecocean/EncounterQueryProcessor.java

```diff
@@ -338,7 +343,132 @@ public static String queryStringBuilder(HttpServletRequest request, StringBuffer
+            }
+          }
+
+        }
+
+        List<String> individualsIds =  new ArrayList<String>(individualsIdsSet);
 
+        String locIDFilter=" ( individual.individualID == \""+individualsIds.get(0)+"\" ";
+        for(int j=1;j<individualsIds.size();j++){
+          locIDFilter+=" || individual.individualID == \""+individualsIds.get(j)+"\" ";
+        }
+        locIDFilter = locIDFilter + " ) ";
+        if(filter.equals(SELECT_FROM_ORG_ECOCEAN_ENCOUNTER_WHERE)){filter+=locIDFilter;}
+        else{filter+=(" && "+locIDFilter);}
+
+        prettyPrint.append("<br />");
+    }
+
+
+      //role search
+       /*
+       * to filter Encounnter based on role fetch all the MarkedIndividuals
+       * for each MarkedIndividuals gets all roles then check if requested role
+       * is not present remove that MarkedIndividuals
+       * finally use these individualIDs to filter the Encounters
+       */
+      if(request.getParameterValues("role")!=null)
+      {
+        boolean orRoles=true;
+
+        Iterator<MarkedIndividual> allSharks = myShepherd.getAllMarkedIndividuals();
+        Vector<MarkedIndividual> rIndividuals=new Vector<MarkedIndividual>();
+        if(allSharks!=null){
+          while (allSharks.hasNext()) {
+            MarkedIndividual temp_shark=allSharks.next();
+            rIndividuals.add(temp_shark);
+          }
+        }
+
+        if(request.getParameter("andRoles")!=null){orRoles=false;}
+        String[] roles=request.getParameterValues("role");
+        int numRoles=roles.length;
+        if(!orRoles){
+          prettyPrint.append("Social roles include all of the following: ");
+        }
+        else{
+          prettyPrint.append("Social roles is one of the following: ");
+        }
+        for(int h=0;h<numRoles;h++){
+          prettyPrint.append(roles[h]+"&nbsp;");
+        }
+
+
+
+        //logical OR the roles
+        for (int q = 0; q<rIndividuals.size(); q++)
+        {
+            MarkedIndividual tShark = (MarkedIndividual) rIndividuals.get(q);
+            List<String> myRoles=myShepherd.getAllRoleNamesForMarkedIndividual(tShark.getIndividualID());
+
+            if(orRoles){
+
+              //logical OR the role
+              boolean hasRole=false;
+              int f=0;
+              while(!hasRole && (f<numRoles)){
+                if(myRoles.contains(roles[f])){hasRole=true;}
+                f++;
+              }
+              if(!hasRole) {
+                  rIndividuals.remove(q);
+                  q--;
+              }
+            }
+            else{
+
+              //logical AND the roles
+              boolean hasRole=true;
+              int f=0;
+              while(hasRole && (f<numRoles)){
+                if(!myRoles.contains(roles[f])){hasRole=false;}
+                f++;
+              }
+              if(!hasRole) {
+                  rIndividuals.remove(q);
+                  q--;
+              }
+
+            }
+
+      }
+
```

Comment:
probably should catch the case here where `rIndividuals.size() == 0` (no individuals matched roles logic), and short-circuit out of there. otherwise the `get(0)` below will throw an exception.

Reply:
thanks for pointing out i have added check for  rIndividuals size is 0  to avoid exception